include/psa/crypto_struct.h

```diff
@@ -68,14 +68,9 @@ extern "C" {
 #include "mbedtls/cipher.h"
 #include "mbedtls/cmac.h"
 #include "mbedtls/gcm.h"
-#include "mbedtls/md.h"
-#include "mbedtls/md2.h"
-#include "mbedtls/md4.h"
-#include "mbedtls/md5.h"
-#include "mbedtls/ripemd160.h"
-#include "mbedtls/sha1.h"
-#include "mbedtls/sha256.h"
-#include "mbedtls/sha512.h"
+
+/* Include the context definition for the compiled-in drivers */
+#include "../../library/psa_crypto_driver_wrappers_contexts.h"
```

Comment:
To use Mbed TLS, it is sufficient to have the headers in `include` and compiled libraries (e.g. `libmbed*.a` or `libmbed*.so` `libmbed*.dll`). In particular, headers in `include` may not depend on headers anywhere else in the source tree.

Reply:
`psa_crypto_driver_wrappers_contexts` needs to move to `include/psa`, yes. But it also needs to lose some things, including references to the test drivers. And yes, drivers need to place their context information under `include`. The driver wrapper generator will produce both `include/psa/something.h` and `library/something.c` as outputs.

Reply:
Test driver contexts are driver contexts, too. So how would `psa_crypto_driver_wrappers_contexts` move to `include/psa` and simultaneously drop the test driver references?

Reply:
Not doing dynamic allocation means all driver contexts need to be fully-typed by the time the driver context union in the struct gets evaluated. So all driver contexts (including the test driver's) need to be reachable from (and thus according to your stated goal, located inside of) the library's `/include` folder.

So... Are we back to creating a `/include/psa/drivers/` folder where all drivers (including both mbedTLS software-based and test) drivers need to put their context structure declarations? And the autogenerated `include/psa/psa_crypto_driver_wrappers_contexts.h` will pull in all declared drivers' context structure headers?

Reply:
Namespacing drivers as `psa/drivers/*.h` seems sensible to me. Mbed TLS itself wouldn't include a `include/psa/drivers` directory though, since we'd have nothing to put there. (Unless we put headers for the built-in implementations there?) If you package Mbed TLS with your own drivers, is it up to Mbed TLS to tell you where to put your headers?

> And the autogenerated include/psa/psa_crypto_driver_wrappers_contexts.h will pull in all declared drivers' context structure headers?

Yes.

Reply:
> Unless we put headers for the built-in implementations there

That's exactly what I was saying. They'd need to go there, since apparently all of their structures need to be declared in `/include`, and just dumping them flat in `include/psa` will (IMO) be terribly messy.